lightning/impl/sag_fast.pyx

```diff
@@ -12,13 +13,86 @@ cimport numpy as np
 
 ctypedef np.int64_t LONG
 
-from libc.math cimport sqrt
+from libc.math cimport sqrt, fabs, fmax
 
 from lightning.impl.randomkit.random_fast cimport RandomState
 from lightning.impl.dataset_fast cimport RowDataset
 from lightning.impl.sgd_fast cimport LossFunction
 
 
+cdef class Penalty:
+
+    cdef void projection(self,
+                         double* w,
+                         int* indices,
+                         double stepsize,
+                         double w_scale,
+                         int n_nz):
+        raise NotImplementedError()
+    
+    cdef void projection_lagged(self,
+                                double* w,
+                                int* indices,
+                                double stepsize,
+                                double* w_scale,
+                                double* scale_cumm,
+                                int t,
+                                int n_nz,
+                                int* last):
+        raise NotImplementedError()
+
+    cdef double regularization(self, np.ndarray[double, ndim=1] coef):
+        raise NotImplementedError()
+
+
+cdef class L1Penalty(Penalty):
+
+
+    cdef void projection(self,
+                         double* w,
+                         int* indices,
+                         double stepsize,
+                         double w_scale,
+                         int n_nz):
+
+        cdef int j, jj
+
+        for jj in xrange(n_nz):
+            j = indices[jj]
+            w[j] = fmax(w[j] - stepsize * w_scale, 0) \
+                    - fmax(-w[j] - stepsize * w_scale, 0)
+
+    cdef void projection_lagged(self,
+                                double* w,
+                                int* indices,
+                                double stepsize,
+                                double* w_scale,
```

Comment:
It does, the scale is used through scale_cum. The w_scale is maintained in case some prox wants to overwrite it.